### PR TITLE
Serve .patch for pull requests

### LINKED ISSUE #3259: Web hook pull notifications .diff and .patch point to different pull and are not diffs
- Gitea version (or commit ref):
- Git version: 1.3.0dev (also see similar issue on 1.3.2)
- Operating system: Debian 7
- Database (use `[x]`):
  - [x ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## Description
 When a webhook for pull happens, it posts:

These 3 urls (pull number always same of course

>     "html_url": "https://try.gitea.io/robe2/test/pulls/1",
    "diff_url": "https://try.gitea.io/robe2/test/pulls/1.diff",
    "patch_url": "https://try.gitea.io/robe2/test/pulls/1.patch"

For try.gitea.io, I can't replicate the diff and patch pull pointing to a different pull, but the urls offered for diff and patch are not diff and patches , they are just the pull html page.

In my production environment these end up pointing to the wrong pull even though they do have the right pull number in the web hook post.  I suspect it might be because the pull it ends up pointing to has the repo deleted.
...

### PR DESCRIPTION
**Pull request by strk**
Closes #3259

### CODE REVIEW

**Comment by strk** [comment]
I don't know why the patch is served as type octect-stream instead of text/plain, ideas @bkcsoft ?

File: routers/repo/pull.go [hunk 1]
```diff
@@ -1054,3 +1055,44 @@ func DownloadPullDiff(ctx *context.Context) {
...
+	if err != nil {
+		if models.IsErrIssueNotExist(err) {
+			ctx.Handle(404, "GetIssueByIndex", err)
+		} else {
+			ctx.Handle(500, "GetIssueByIndex", err)
+		}
+		return
+	}
+
+	// Redirect elsewhere if it's not a pull request
```
Comment (lafriks) [inline routers/repo/pull.go]: Wrong comment

**Review by lafriks — APPROVED** [review]

**Comment by lunny** [comment]
@strk I would like this based on https://github.com/go-gitea/git/pull/105

File: integrations/pull_create_test.go [hunk 2]
```diff
@@ -59,4 +59,12 @@ func TestPullCreate(t *testing.T) {
 	assert.Regexp(t, "\\+Hello, World \\(Edited\\)", resp.Body)
 	assert.Regexp(t, "^diff", resp.Body)
 	assert.NotRegexp(t, "diff.*diff", resp.Body) // not two diffs, just one
+
+	// check .patch can be accessed and matches performed change
+	req = NewRequest(t, "GET", url+".patch")
+	resp = session.MakeRequest(t, req, http.StatusOK)
+	assert.Regexp(t, "\\+Hello, World \\(Edited\\)", resp.Body)
```
Comment (thehowl) [inline integrations/pull_create_test.go]: Why not <code>\`\\+Hello, World \\(Edited\\)\`</code> (raw string)? Seems cleaner to me

File: integrations/pull_create_test.go [hunk 3]
```diff
@@ -59,4 +59,12 @@ func TestPullCreate(t *testing.T) {
...
+	assert.Regexp(t, "diff", resp.Body)
+	assert.Regexp(t, "Subject: \\[PATCH\\] Update 'README.md'", resp.Body)
```
Comment (thehowl) [inline integrations/pull_create_test.go]: Same here.

**Review by thehowl — APPROVED** [review]
Two minor formatting nitpicks, otherwise looking good

**Comment by strk** [comment]
I didn't understand the suggestion of raw string, are you asking
why I'm using regexp instead of substring ?

**Comment by lafriks** [comment]
@strk he was suggesting to use golang raw strings to not to have escape backslash characters in it

**Comment by strk** [comment]
Use of raw strings pushed. As per @lunny PR I'll wait for it to be merged before changing this one.
As usual, I'd rather not block things, and if @lunny *adds to*, rather than *change* the code, nothing will break...

**Comment by strk** [comment]
I've updated to new `git` method and squash-rebased to master.
The updated code fixes inlinling of patch into browser, so ready to merge for me.

**Comment by lunny** [comment]
@strk you should handle `io.Copy` returned error.

**Comment by strk** [comment]
error handled with 8243e4f4

File: routers/repo/pull.go [hunk 4]
```diff
@@ -1054,3 +1055,48 @@ func DownloadPullDiff(ctx *context.Context) {
...
+		return
+	}
+
+	patch, err := headGitRepo.GetFormatPatch(pr.MergeBase, pr.HeadBranch)
+	if err != nil {
+		ctx.Handle(500, "GetFormatPatch", err)
+		return
+	}
+
+	_, err = io.Copy(ctx, patch)
```
Comment (bkcsoft) [inline routers/repo/pull.go:1097]: @strk You need to set the approriate `Content-Type` before this :)

Comment (strk) [reply routers/repo/pull.go:1097]: The PR is merged, if you can handle to test Content-Type in the existing integration test that'd be very useful (Content-Type was automatically added in previous implementation via ServeContent)

**Pull request merged**